### PR TITLE
Use golangci-lint-action instead of `make golangci-lint` for main.yml workflow

### PR DESCRIPTION
**Pull request by arkid15r**
### Describe Your Changes

This PR splits `main.yml` lint job `Run check-all` step into multiple steps and switches to [golangci-lint action](https://github.com/golangci/golangci-lint-action) (see the [Performance](https://github.com/golangci/golangci-lint-action?tab=readme-ov-file#performance) section)

### Checklist

The following checks are **mandatory**:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

**Review by zekker6 — COMMENTED** [review]
Could you try replacing golangci-lint with an official action to see the difference in performance?
Based on current actions run it took 32s for golangcilint alone.

**Comment by arkid15r** [comment]
> Could you try replacing golangci-lint with an official action to see the difference in performance? Based on current actions run it took 32s for golangcilint alone.

Yes, the eventual goal is to keep the most efficient option. I'd like to get some more data before making the switch decision. This PR's goal is to extract the `golangci-lint` step to see its "clean" running time. I was planning  to implement the suggestion you've mentioned in 1-2 days when we have several `main` workflow runs.

**Review by zekker6 — APPROVED** [review]

**Comment by hagen1778** [comment]
@arkid15r @zekker6 can we do this experiment in fork? Or does it require paid subscription?

**Comment by zekker6** [comment]
@hagen1778 As far as I can see, @arkid15r wants to gather statistics about real usage before updating the github action. If that is not the case, I don't see any blockers for testing in fork.

**Comment by hagen1778** [comment]
> @hagen1778 As far as I can see, @arkid15r wants to gather statistics about real usage before updating the github action. If that is not the case, I don't see any blockers for testing in fork.

The time spent by golangci-lint will be the same for the fork. And if not, if there are caches that reduce this time, you always can make a PR to fork touching as many go files as you want.

**Comment by arkid15r** [comment]
I've updated the PR based on my fork repo job run results (make golangci-lint running time: 44s, 31s, 30s; golangci-lint action: 30s first run, 4s, 6s, 3s). 

@hagen1778 PTAL

**Comment by hagen1778** [comment]
@arkid15r @zekker6 please correct if my understanding of this PR is wrong:
1. `golangci/golangci-lint-action@v6` is faster than `golangci-lint run`
2. We want to switch to `golangci/golangci-lint-action@v6` to save the CI time

If this understanding is correct, then I have the following questions:
1. Are `golangci/golangci-lint-action@v6` and `golangci-lint` interchangeable? It is important to be able to reproduce the check locally
2. Will `- name: Check diff` as a separate action capture stdout of previous actions? Especially `name: Run golangci-lint` action?

**Comment by zekker6** [comment]
@hagen1778 
> Are golangci/golangci-lint-action@v6 and golangci-lint interchangeable? It is important to be able to reproduce the check locally

Yes, this action is a wrapper for the same golangci-lint project and uses the same binary. The benefit of using an action is additional caching it provides, such as cache of the installed binary and scanned files. 

> Will - name: Check diff as a separate action capture stdout of previous actions? Especially name: Run golangci-lint action?

`git diff --exit-code` should print changes between local state of the code and commited version. I assume it was added in order to print changes made by `go vet` / `go fmt` (if there are any) so that it will be easier to apply suggested fixes.
Since all actions are using the same FS with clone repo it will be able to provide the same output as it did before.

**Comment by arkid15r** [comment]
It pretty much aligns with my understanding of the workflow. A couple points to clarify:

> Will - name: Check diff as a separate action capture stdout of previous actions? Especially name: Run golangci-lint action?

It shouldn't depend on (technical) stdout of previous steps but rather on the git filesystem state.


> in order to print changes made by go vet / go fmt (if there are any) so that it will be easier to apply suggested fixes.

I guess the reason is not only print it but also fail the workflow job explicitly if lint process produces any changes.

As @zekker6 stated the outcome expected to be exactly the same as before the refactoriing. I only wasn't sure if it'd make sense to move `git diff --exit-code` step to better place (I'm not sure if `govulncheck` produces any git diff change).

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -46,10 +48,24 @@ jobs:
...
+
+      - name: Run vet
+        run: make vet
+      
+        # Use action instead of `make golangci-lint` to speed up the process
+        # as it caches data between builds.
+      - name: Run golangci-lint
+        uses: golangci/golangci-lint-action@v6
+        with:
+          version: v1.59
```
Comment (f41gh7) [inline .github/workflows/main.yml:62]: this change introduces discrepancy for local and CI commands.

golangci-lint version requires to be in sync with Makefile and github ci. As far as I remember we already had such problem and decided to use only make file commands https://github.com/VictoriaMetrics/VictoriaMetrics/commit/c560a338e8c41b1854b1e0bdb4ce2697ddf68956

Comment (arkid15r) [reply .github/workflows/main.yml]: I've fixed the mismatch and added comments in each part to remind keeping it in sync.
I guess it's something we can live with as a tradeoff for more performant lint process? Let me know if I'm wrong.

**Comment by hagen1778** [comment]
Hey @arkid15r! I appreciate the work and tests done in this PR. But I think I'd agree with @f41gh7 point - local and CI commands must be the same. Otherwise, it could potentially lead to confusion, which would be hard to understand and address for engineers. I'm closing this PR as won't be accepted.

**Pull request closed without merging**